Enable transcription of native commands on non-Windows

SteveL-MSFT opened the pull request:
Transcription was relying on reading the screen buffer to record output from native commands.
This resulted in an unhandled exception calling an unimplemented api on non-Windows.
Fix is to default to using redirected output if reading the screen buffer is not supported.

Fix https://github.com/PowerShell/PowerShell/issues/1920

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md).

-->

SteveL-MSFT committed:
handle exception for non-Windows where reading the screen buffer is not supported
since screen buffer reading isn't supported, default to using redirected output

anmenaga commented on test/powershell/Modules/Microsoft.Powershell.Host/Start-Transcript.Tests.ps1:
```diff
@@ -136,4 +135,19 @@ Describe "Start-Transcript, Stop-Transcript tests" -tags "CI" {
+                $ps.Dispose()
+            }
+        }
+        Test-Path $transcriptFilePath | Should be $true
```
Also need to check that PS error stream is empty.

SteveL-MSFT replied:
will add

SteveL-MSFT replied:
Actually, there's no reason to follow that pattern, I'm rewriting it as a scriptblock and any errors should just show up in the test run.

anmenaga commented on test/powershell/Modules/Microsoft.Powershell.Host/Start-Transcript.Tests.ps1:
```diff
@@ -136,4 +135,19 @@ Describe "Start-Transcript, Stop-Transcript tests" -tags "CI" {
+        $machineName = [System.Environment]::MachineName
+        $transcriptFilePath | Should contain $machineName
+    }
 }
```
Need newline in the end of file.

SteveL-MSFT replied:
will add

SteveL-MSFT committed:
address PR feedback

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -185,6 +185,17 @@ internal NativeCommandProcessor(ApplicationInfo applicationInfo, ExecutionContex
+            {
+                Host.BufferCell[,] bufferContents = this.Command.Context.EngineHostInterface.UI.RawUI.GetBufferContents(
+                    new Host.Rectangle(_startPosition, _startPosition));
+                _scrapeHostOutput = true;
```
This can be made a static check so that we do it only once. And `_scrapeHostOutput` can be renamed to `s_supportScreenScrape` (or something similar).

SteveL-MSFT replied:
will change

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -398,12 +409,16 @@ private void InitNativeProcess()
+            // Calculate if input and output are redirected.
+            bool redirectOutput = true;
+            bool redirectError = true;
+            bool redirectInput = true;
```
Why do we want `redirectInput` to be true by default? Even if we are transcribing and screen scrape is not supported, we shouldn't need to always redirect the input.

daxian-dbw replied:
And form line 442 below, it indicates that we transcribe for native command only if `_runStandAlone` is true. Now it looks like you are assuming we can do transcribe regardless of `_runStandAlone`. That seems not right.

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -398,12 +409,16 @@ private void InitNativeProcess()
 
-            CalculateIORedirection(out redirectOutput, out redirectError, out redirectInput);
+            // If we are transcribing and can't scrape, then continue to redirect so we have the output
+            if (!this.Command.Context.EngineHostInterface.UI.IsTranscribing || _scrapeHostOutput)
```
Suggestion: `this.Command.Context.EngineHostInterface.UI.IsTranscribing` is used at 3 places in this class. Maybe we can have an instance field to cache the value? Like `_isTranscribing = this.Command.Context.EngineHostInterface.UI.IsTranscribing` in the constructor?

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -432,19 +446,10 @@ private void InitNativeProcess()
-                            _startPosition.X = 0;
-                        }
-                    }
-                    catch (Host.HostException)
```
Looks like an HostException also indicates `The host doesn't support scraping via its RawUI interface` ...

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -185,6 +185,17 @@ internal NativeCommandProcessor(ApplicationInfo applicationInfo, ExecutionContex
+            catch (NotImplementedException)
+            {
+                _scrapeHostOutput = false;
+            }
```
from line 444 below, it looks like a `HostException` also indicates `The host doesn't support scraping via its RawUI interface` ...

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -185,6 +185,17 @@ internal NativeCommandProcessor(ApplicationInfo applicationInfo, ExecutionContex
+                    new Host.Rectangle(_startPosition, _startPosition));
+                _scrapeHostOutput = true;
+            }
+            catch (NotImplementedException)
```
By putting the call to `GetBufferContents` in the constructor, it's possible that an arbitrary exception will be thrown from `GetBufferContents` in the constructor. When that happens, it would be wrapped to an `ApplicationFailedException` properly prior to this change, but after this change, the `NativeCommandProcessor` would just fail to be constructed, and thus I guess the error would look more obscure. So, I'm not sure if it's appropriate to put this in the constructor ...

SteveL-MSFT replied:
will fix

daxian-dbw commented:
> Fix is to default to using redirected output if reading the screen buffer is not supported.

@SteveL-MSFT Could you please explain a little why redirecting the output/error would support transcribing in case that screen scraping is not supported? (just curious how it's done in that case)

SteveL-MSFT commented:
@daxian-dbw from what I can tell in the code, anything that gets redirected gets transcribed anyways

SteveL-MSFT committed:
address Dongbo's PR feedback

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -698,8 +708,7 @@ internal override void Complete()
                     // Capture screen output if we are transcribing
-                    if (this.Command.Context.EngineHostInterface.UI.IsTranscribing &&
-                        _scrapeHostOutput)
+                    if (_isTranscribing && (true == s_supportScreenScrape))
```
I think there is a change of behavior here. Assuming `_isTranscribing == $true` and screen scraping is supported,
- before the change, `_scrapeHostOutput` is true only if `_runStandAlone` is true, so this block will run only the native command is running standalone.
- after the change, even if it's not running standalone (i.e. standard input/error redirected), this block of code still will run when transcribing, and in such case, `_startPosition` is not even initialized.

SteveL-MSFT replied:
Will fix

SteveL-MSFT commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -398,12 +400,30 @@ private void InitNativeProcess()
+            }
+
+            // If we are transcribing and can't scrape, then continue to redirect so we have the output
+            if (!_isTranscribing || (true == s_supportScreenScrape))
```
Makes sense, will fix.

SteveL-MSFT committed:
[feature]
moved setting redirection due to transcription to within CalculateIORedirection() function
ensure screen scrapping only happens when running standalone

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -1285,6 +1290,11 @@ private void CalculateIORedirection(out bool redirectOutput, out bool redirectEr
                     redirectError = true;
                 }
             }
+            else if (_isTranscribing && (false == s_supportScreenScrape))
```
Maybe remove the `else`? The `else if (Platform.IsWindowsDesktop && IsConsoleApplication(this.Path))` block doesn't necessarily set `redirectOutput` and `redirectError` to true.

daxian-dbw replied:
And please add a comment that explains why we need to set output/error redirection in such case.

SteveL-MSFT replied:
yup

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -398,10 +400,26 @@ private void InitNativeProcess()
+            // Calculate if input and output are redirected.
+            bool redirectOutput = true;
+            bool redirectError = true;
+            bool redirectInput = this.Command.MyInvocation.ExpectingInput;
```
These changes now can be reverted.

SteveL-MSFT replied:
Will revert

SteveL-MSFT committed:
[feature]
address PR feedback from Dongbo

daxian-dbw commented:
@anmenaga Can you please take a look again?

SteveL-MSFT commented:
@markekraus the test failures are indicating the WebListener didn't start within the timeout.  Is this something you're familiar with?

markekraus commented:
@SteveL-MSFT News to me. I will take a look.

markekraus commented:
@SteveL-MSFT This appears to be related to the change in this PR. It looks native execution inside a job is broken. WebListener runs `dotnet WebListener.dll` in a `Start-Job` block. From your PR I can manually run the WebListener outside of a job in PowerShell. But inside a job it fails. But it's not limited to WebListener:

```poweshell
$Job = Start-Job {cmd.exe /c echo "test" }
$Job | Wait-Job | fl * -force
"-{0}-" -f $Job.ChildJobs[0].Output
```

expected behavior:
```none
State         : Completed
HasMoreData   : True
StatusMessage :
Location      : localhost
Command       : cmd.exe /c echo "test"
JobStateInfo  : Completed
Finished      : System.Threading.ManualResetEvent
InstanceId    : 7d654dd9-ba42-48f0-be72-9e91863e7286
Id            : 1
Name          : Job1
ChildJobs     : {Job2}
PSBeginTime   : 9/26/2017 3:41:02 PM
PSEndTime     : 9/26/2017 3:41:02 PM
PSJobTypeName : BackgroundJob
Output        : {}
Error         : {}
Progress      : {}
Verbose       : {}
Debug         : {}
Warning       : {}
Information   : {}

-test-
```

Actual behavior:
```none
State         : Failed
HasMoreData   : False
StatusMessage :
Location      : localhost
Command       : cmd.exe /c echo "test"
JobStateInfo  : Failed
Finished      : System.Threading.ManualResetEvent
InstanceId    : 58c30632-c4a1-41e3-b276-5aa9b24db3b0
Id            : 3
Name          : Job3
ChildJobs     : {Job4}
PSBeginTime   : 9/26/2017 3:42:41 PM
PSEndTime     : 9/26/2017 3:42:42 PM
PSJobTypeName : BackgroundJob
Output        : {}
Error         : {}
Progress      : {}
Verbose       : {}
Debug         : {}
Warning       : {}
Information   : {}

--
```

SteveL-MSFT commented:
@markekraus thanks!  I'll take a look.

SteveL-MSFT committed:
[feature]
added test for start-job and native commands
catch all exceptions trying to read the buffer

daxian-dbw commented on src/System.Management.Automation/engine/NativeCommandProcessor.cs:
```diff
@@ -398,11 +400,25 @@ private void InitNativeProcess()
+            catch (Exception)
+            {
+                // screen scraping not supported on non-Windows or as job
+            }
```
I think this check can still be done only once -- we do this check in `CalculateIORedirection` and only if `_runStandAlone` is true. Here the the code I propose:
```CSharp
private void CalculateIORedirection(out bool redirectOutput, out bool redirectError, out bool redirectInput)
{
    ...

    if (NativeCommandProcessor.IsServerSide)
    {
        redirectInput = true;
        redirectOutput = true;
        redirectError = true;
    }
    else if (Platform.IsWindowsDesktop && IsConsoleApplication(this.Path))
    {
        // On Windows desktops, if the command to run is a console application,
        // then allocate a console if there isn't one attached already...
        ConsoleVisibility.AllocateHiddenConsole();

        if (ConsoleVisibility.AlwaysCaptureApplicationIO)
        {
            redirectOutput = true;
            redirectError = true;
        }
    }
    _runStandAlone = !redirectInput && !redirectOutput && !redirectError;

    // !!!!! NOTE !!!!! -- unchanged until here

    if (_runStandAlone)
    {
        if (s_supportScreenScrape == null)
        {
            // Call GetBufferContents() to check and set s_supportScreenScrape to true or false accordingly.
        }

        // if screen scraping isn't supported, we enable redirection so that the output is still transcribed
        // as redirected output is always transcribed
        if (_isTranscribing && (false == s_supportScreenScrape))
        {
            redirectOutput = true;
            redirectError = true;
            _runStandAlone = false;
        }
    }
    
```
So the check of screen scraping support is done only if we are running the native command as standalone application.

In case of running it in a Job, I think the standard input and error are already redirected.

SteveL-MSFT replied:
I see.  I'll try that.

SteveL-MSFT committed:
[feature]
refactor code

daxian-dbw reviewed (approved):
LGTM. Thanks!
